7882 add message tracking in entity emailer to fix duplicate email issue

argush3 opened the pull request:
*Issue #:* /bcgov/entity#7882

*Description of changes:*

* Addition of new database with a `message_processing` table used to track processing of messages
* Add service and utility classes to encapsulate message tracking logic
* Integrate message tracking into entity-emailer `worker.py`

![image](https://user-images.githubusercontent.com/6685223/123285985-5af9db00-d4c2-11eb-86d0-d8d32fe9fdfc.png)

![image](https://user-images.githubusercontent.com/6685223/123286004-61885280-d4c2-11eb-91c6-1f1b080252f7.png)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

argush3 commented on queue_services/entity-emailer/src/entity_emailer/email_tracker/tracker.py:
```diff
@@ -0,0 +1,131 @@
+from tracker.services import MessageProcessingService
+
+
+def get_message_id(raw_queue_msg: nats.aio.client.Msg):
```
would like to just use the cloud event id to uniquely identify a payload but not all publishers use the cloud event format yet.  Needed to implement custom logic to uniquely identify a queue message id on a case to case basis.

argush3 commented on queue_services/entity-emailer/src/entity_emailer/config.py:
```diff
@@ -97,6 +97,23 @@ class _Config():  # pylint: disable=too-few-public-methods
+        name=TRACKER_DB_NAME,
+    )
+
+    SQLALCHEMY_BINDS = {
```
This was required in order to support binding to lear db and tracker db.  Model classes for tracker db require a ` __bind_key__ = 'tracker'` in order to support multiple db binds.

argush3 commented on queue_services/entity-emailer/tracker/models/message_processing.py:
```diff
@@ -0,0 +1,68 @@
+
+    """
+
+    __bind_key__ = 'tracker'
```
Required to support multiple db binds

pwei1018 commented on tracker-db/Jenkinsfile:
```diff
@@ -0,0 +1,99 @@
+// Edit your app's name below
```
Don't need the Jenkinfile.

argush3 replied:
removed

pwei1018 commented on tracker-db/Dockerfile:
```diff
@@ -0,0 +1,5 @@
+FROM postgres:11
```
Could change to version 12.

argush3 replied:
updated to 12

pwei1018 commented on tracker-db/Jenkinsfile.pipeline.param:
```diff
@@ -0,0 +1,11 @@
+#=========================================================
```
Don't need it.

argush3 replied:
removed

pwei1018 commented on tracker-db/backup/backup.yaml:
```diff
@@ -0,0 +1,284 @@
+---
+kind: Template
```
Don't need this backup. The lear db backup can back this db as well.

argush3 replied:
removed

pwei1018 commented on tracker-db/openshift/templates/postgresql-build.json:
```diff
@@ -0,0 +1,119 @@
+{
```
Don't need this.

argush3 replied:
removed

pwei1018 commented on tracker-db/openshift/postgresql-deploy.tracker.test.param:
```diff
@@ -0,0 +1,20 @@
+#=========================================================
```
Don't need this.

pwei1018 commented on tracker-db/openshift/postgresql-deploy.tracker.prod.param:
```diff
@@ -0,0 +1,20 @@
+#=========================================================
```
Don't need this.

pwei1018 commented on tracker-db/openshift/postgresql-deploy.tracker.dev.param:
```diff
@@ -0,0 +1,20 @@
+#=========================================================
```
Don't need this.

argush3 replied:
removed

pwei1018 commented on tracker-db/openshift/postgresql-build.tracker.param:
```diff
@@ -0,0 +1,13 @@
+#=========================================================
```
We don't use this kind of param files any more.

argush3 replied:
removed

pwei1018 commented on tracker-db/openshift/templates/postgresql-deploy.json:
```diff
@@ -0,0 +1,359 @@
+{
```
Don't need this.

argush3 replied:
removed

pwei1018 commented on tracker-db/openshift/templates/tracker-db.yaml:
```diff
@@ -0,0 +1,270 @@
+    displayName: Persistent Volume Size
+    description: The size of the persistent volume , e.g. 512Mi, 1Gi, 2Gi.
+    required: true
+    value: 10Gi
```
Need to review the size of PVC. The namespace may not have enough space for it.

argush3 replied:
Need confirmation on what value should be used for this.  Have left as is for now.

pwei1018 replied:
Currently, lear db and backup requested total  30G space. It has 20G leave over. You may need to estimate your database size for db and backup.

argush3 replied:
hmmmm...it's only one table but it will grow over time as it tracks completed messages too.  Ideally we would remove completed messages after a certain threshold.  Let's say we start with 15GB for tracker_db and its backup.  If that's the case, do you need to do anything additional to support this or can I just update to 15Gi in the above yaml section?

argush3 replied:
As per our conversation, this has now been updated to 5G.  

The database is relatively small and inserting 1000 test records came out to be 560kB.  Extrapolating that to 2 years gives us 409 MBs.

pwei1018 replied:
cool

pwei1018 commented:
The tracker-db folder only need a README.md, tracker-db.yaml, sql and cfg. Other files we don’t it.

thorwolpert commented on queue_services/entity-emailer/tracker/app.py:
```diff
@@ -0,0 +1,45 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+"""
```
why is this blank?

argush3 replied:
My bad.  Updated to include description.

thorwolpert commented on queue_services/entity-emailer/tracker/config.py:
```diff
@@ -0,0 +1,113 @@
+# Copyright © 2019 Province of British Columbia
```
did you look at using the config from the root level and just using the bind here?

argush3 replied:
I looked at this a bit more.  I was able to figure out a way to use the root config file in the end.  It involves initializing the alembic stuff with `python manage.py db init --multidb`.  Doing that generates a 'env.py` in the migrations folder that looks at `SQLALCHEMY_BINDS`.  I did need to comment out some code in `env.py` to prevent migration script generation functionality from trying to run examine lear db.  And the migration scripts also contain more code to handle multiple databases.  It does work it seems.  I haven't put this code for now as I'm still a little hesitant there might be paths where alembic is trying to migrate/upgrade the lear db.

thorwolpert replied:
cool, that sounds nice.
i wouldn't put too much more time on that. when we tackle putting this into the common services, we can finalize ... and it looks like we may leverage what you've learned here.

thorwolpert reviewed (commented):
Probably need to discuss this, as it looks like this will now have multiple FLASK instances.

argush3 commented:
@thorwolpert added a `message_seen` and `last_error` column to aid ops as you had mentioned.  

Also in digging into the CloudEvent spec context attributes a little more deeply, I realized that the `source` + `id` is what the CloudEvent spec uses to uniquely identify an event.  As such, I added a `source` column and updated the code to look for `source` + `id` to uniquely identify a message when a message is using the CloudEvent format.
